core-tests/shared/src/test/scala/zio/ServiceProxySpec.scala

```diff
@@ -0,0 +1,270 @@
+        val service: Foo = new Foo { def bar(a: Int)(b: String) = ZIO.succeed(b * a) }
+        for {
+          ref  <- ScopedRef.make(service)
+          proxy = ServiceProxy.generate(ref)
+          res  <- proxy.bar(3)("zio")
+        } yield assertTrue(res == "zioziozio")
+      },
+      test("implicit clauses") {
+        trait Foo { def bar(a: Int)(implicit b: String): UIO[String] }
+        val service: Foo       = new Foo { def bar(a: Int)(implicit b: String) = ZIO.succeed(b * a) }
+        implicit val b: String = "zio"
+        for {
+          ref  <- ScopedRef.make(service)
+          proxy = ServiceProxy.generate(ref)
+          res  <- proxy.bar(3)
+        } yield assertTrue(res == "zioziozio")
+      },
+      test("inherited abstract methods") {
+        trait Foo0 { def bar(a: Int): UIO[String] }
+        trait Foo extends Foo0
+
+        val service: Foo = new Foo { def bar(a: Int) = ZIO.succeed("zio" * a) }
+        for {
+          ref  <- ScopedRef.make(service)
+          proxy = ServiceProxy.generate(ref)
+          res  <- proxy.bar(3)
+        } yield assertTrue(res == "zioziozio")
+      },
+      test("overridden methods with default implementation") {
+        trait Foo { def bar: UIO[String] = ZIO.succeed("zio1") }
+        val service: Foo = new Foo { override def bar = ZIO.succeed("zio2") }
+        for {
+          ref  <- ScopedRef.make(service)
+          proxy = ServiceProxy.generate(ref)
+          res  <- proxy.bar
+        } yield assertTrue(res == "zio2")
+      },
+      test("package private methods") {
+        trait Foo { private[zio] def bar: UIO[String] }
+        val service: Foo = new Foo { private[zio] def bar: UIO[String] = ZIO.succeed("zio") }
+        for {
+          ref  <- ScopedRef.make(service)
+          proxy = ServiceProxy.generate(ref)
+          res  <- proxy.bar
+        } yield assertTrue(res == "zio")
+      },
+      test("ZIO vals") {
+        trait Foo { val bar: UIO[String] }
+        val service1: Foo = new Foo { val bar: UIO[String] = ZIO.succeed("zio1") }
+        val service2: Foo = new Foo { val bar: UIO[String] = ZIO.succeed("zio2") }
+        for {
+          ref  <- ScopedRef.make(service1)
+          proxy = ServiceProxy.generate(ref)
+          res1 <- proxy.bar
+          _    <- ref.set(ZIO.succeed(service2))
+          res2 <- proxy.bar
+        } yield assertTrue(res1 == "zio1" && res2 == "zio2")
+      },
+      test("ZIO vals with default implementation") {
+        trait Foo { val bar: UIO[String] = ZIO.succeed("zio") }
+        val service: Foo = new Foo { override val bar: UIO[String] = ZIO.succeed("zio1") }
+        for {
+          ref  <- ScopedRef.make(service)
+          proxy = ServiceProxy.generate(ref)
+          res1 <- proxy.bar
+        } yield assertTrue(res1 == "zio1")
+      },
+      test("keeps non-ZIO default implementations") {
+        trait Foo {
+          def bar: UIO[String]
+          def qux: String = "quux"
```

Comment:
I think this is correct behavior for things like constants. Could be extra strict and require they be `final` for bulletproof semantics.